File: array/898.子数组按位或操作.cpp
/*
* @lc app=leetcode.cn id=898 lang=cpp
*
* [898] 子数组按位或操作
*/

#include"head.h"
using namespace std;
// @lc code=start
class Solution{
    public:
int subarrayBitwiseORs(vector<int>& A) {
	int maxval = *max_element(A.begin(), A.end());
	int mask = 1;
	while (maxval >= mask)
		mask <<= 1;
	mask -= 1;
	unordered_set<int> nums;
	for (int i = 0; i < A.size(); ++i) {
		int val = A[i];
		nums.insert(val);
		for (int j = i - 1; j >= 0 && val != mask; --j) {
			val |= A[j];
			nums.insert(val);
		}
	}
	return nums.size();
}
};
// @lc code=end

File: array/898.子数组按位或操作.go
package main

/*
 * @lc app=leetcode.cn id=898 lang=golang
 *
 * [898] 子数组按位或操作
 */

// @lc code=start
func subarrayBitwiseORs(arr []int) int {
    set := make(map[int]bool)
    // set[arr[0]] = true
    for i:=0;i<len(arr);i++ {
        temp := map[int]bool{arr[i]:true}
        x := arr[i]
        for j :=i-1;j>=0;j-- {
            if  x | arr[j] == arr[j] {
                break
            }
            arr[j] = x | arr[j]
            temp[arr[j]] = true;
        }
        for k,_ := range temp {
            set[k] = true
        }
    }
    return len(set)
    
}
// @lc code=end

